crates/sui-core/src/transaction_driver/effects_certifier.rs

```diff
@@ -269,6 +302,98 @@ impl EffectsCertifier {
         }
     }
 
+    /// Gets full effects from a validator, with speculative fallback to other validators.
+    ///
+    /// If the initial validator doesn't respond within GET_FULL_EFFECTS_FALLBACK_DELAY,
+    /// we start parallel requests to validators that have already acknowledged the effects
+    /// (received via the acked_validators channel from wait_for_acknowledgments).
+    ///
+    /// This prevents slow validators from blocking the entire operation when faster
+    /// validators are available, while still preferring the initial validator if it responds quickly.
+    #[instrument(level = "debug", skip_all, fields(tx_digest = ?tx_digest, initial_validator = ?initial_target))]
+    async fn get_full_effects_with_fallback<A>(
+        &self,
+        authority_aggregator: &Arc<AuthorityAggregator<A>>,
+        initial_client: Arc<SafeClient<A>>,
+        initial_target: AuthorityName,
+        tx_digest: Option<TransactionDigest>,
+        tx_type: TxType,
+        consensus_position: Option<ConsensusPosition>,
+        options: &SubmitTransactionOptions,
+        mut acked_validators_rx: Receiver<AuthorityName>,
+    ) -> (FullEffectsResult, AuthorityName)
+    where
+        A: AuthorityAPI + Send + Sync + 'static + Clone,
+    {
+        let initial_request = self.get_full_effects(
+            initial_client,
+            tx_digest,
+            tx_type,
+            consensus_position,
+            options,
+        );
+        tokio::pin!(initial_request);
+
+        let mut fallback_delay = tokio::time::interval(GET_FULL_EFFECTS_FALLBACK_DELAY);
+        fallback_delay.tick().await;
+
+        // Track validators we've already sent requests to
+        let mut requested_validators = HashSet::new();
+        requested_validators.insert(initial_target);
```

Comment:
since we assume that each validator is sending at most one ack , then do we need the `HashSet` to dedupe?

Reply:
Yeah actually only need to guard against the initial target as we can get an ack for it and we dont want to resubmit the get_full_effects. Will simplify this so its more clear